Decorators and forwarding, call/apply (#218)
javascript-tutorial/pt.javascript.info: 1-js/06-advanced-functions/09-call-apply-decorators/article.md

alert( "Again: " + slow(1) ); // slow(1) o resultado é retornado da cache

alert( slow(2) ); // slow(2) é cacheada e o resultado retornado
alert( "Again: " + slow(2) ); // slow(2) o resultado é retornado da cache

```

No código acima `cachingDecorator` é um *decorador*: uma função especial que recebe outra função e altera o seu comportamento.

A ideia é que podemos chamar `cachingDecorator` com qualquer função, e ele retornará o encapsulador de cache. É ótimo, porque podemos ter muitas funções que poderiam usar essa funcionalidade, e tudo o que precisamos de fazer é aplicar `cachingDecorator` a elas.

Ao separar o cache do código principal da função, também mantemos o código principal mais simples.

Agora, vamos entrar em detalhes de como isso funciona.

O resultado do `cachingDecorator(func)` é um "encapsulador": `function(x)` que "envolve" a chamada de `func(x)` dentro duma lógica de cache:

![](decorator-makecaching-wrapper.svg)

Como podemos ver, o encapsulador retorna o resultado da `func(x)` "como ele é". A partir de um código exterior, a função `slow` encapsulada ainda faz o mesmo. Ela apenas teve o aspecto de cache adicionado ao seu comportamento.

> Resumindo, existem muitos benefícios em usar um `cacheDecorator` separado ao invés de alterar o próprio  código de `slow`:

Reviewer: ```suggestion
Resumindo, existem muitos benefícios em usar um `cacheDecorator` separado ao invés de alterar o próprio código de `slow`:
```

- O `cacheDecorator` é reusável. Nós o podemos aplicar a outras funções.
- A lógica de cache está separada, ela não adicionou complexidade a `slow` (se havia alguma).
- Nós podemos combinar vários decoradores se necessário (outros decoradores virão a seguir).

## Usando "func.call" para o contexto

> O decorador de cache mencionado acima não é adequado para trabalhar com os métodos de objeto.

Reviewer: ```suggestion
O decorador de cache mencionado acima não é adequado para trabalhar com métodos de objeto.
```

let user = { name: "John" };

> // user se torna this, e "Hello" se torna o primeiro argumento

Reviewer: ```suggestion
// user se torna this, e "Hello" se torna no primeiro argumento
```

worker.slow = cachingDecorator(worker.slow);
```

Temos dois problemas para resolver aqui.

Primeiro, é como usar ambos os argumentos `min` e `max` com uma chave no mapa de `cache`. Anteriormente, para um único argumento `x` poderiamos apenas `cache.set(x, result)` para guardar o resultado e `cache.get(x)` para recuperá-lo. Porém agora precisamos de nos lembrar do resultado para uma *combinação de argumentos* `(min,max)`. O `Map` nativo recebe somente um único valor como chave.

Existem muitas soluções possíveis:

1. Implemente uma nova (ou use a de um terceiro fornecedor) estrutura de dados parecida ao mapa que seja mais versátil e que permita várias chaves.
2. Use mapas aninhados: `cache.set(min)` será um `Map` que armazena o par `(max, result)`. Então, podemos receber `result` como `cache.get(min).get(max)`.
3. Junte dois valores em um. No nosso caso particular podemos apenas usar uma string `"min,max"` como a chave do `Map`. Para flexibilidade, podemos permitir prover uma *função de hashing* ao decorador, que saiba como fazer um valor a partir de muitos.

Para muitas aplicações práticas, a terceira variante é suficientemente boa, então vamos segui-la.

> Também precisamos de passar não apenas o `x`, mas todos argumentos em `func.call`. Vamos recapitular que em uma `function()` podemos receber um pseudo-array de seus argumentos como `arguments`, então `func.call(this, x)` deveria ser substituida por `func.call(this, ...arguments)`.

Reviewer: ```suggestion
Também precisamos de passar não apenas o `x`, mas todos argumentos em `func.call`. Vamos recapitular que em uma `function()` podemos receber um pseudo-array com os seus argumentos como `arguments`, então `func.call(this, x)` deveria ser substituida por `func.call(this, ...arguments)`.
```

```

Ele executa `func` configurando `this=context` e usando um objeto parecido com  array `args`  como lista de argumentos.

A única diferença de sintáxe entre `call` e `apply` é que `call` espera uma lista de argumentos, enquanto `apply` recebe um objeto parecido com um array com eles.

Então essas duas chamadas são quase equivalentes:

```js
func.call(context, ...args);
func.apply(context, args);
```

Eles executam a mesma chamada de `func` com o contexto e argumentos dados.

Há somente uma diferença sútil em relação a `args`:

- A sintaxe spread (espalhar) `...` permite passar `args` *iterável* como a lista para `call`.
- O `apply` aceita somente `args` *parecido com array (array-like)*.

...E para objetos que são ambos iteráveis e parecidos com array, tal como um verdadeiro array, podemos usar qualquer um deles, porém `apply` irá provavelmente ser o mais rápido, porque a maioria dos interpretadores de JavaScript internamente otimizam ele melhor.

Passar todos argumentos junto com o contexto para outra função é chamado de *encaminhamento de chamada*.

Esta é a forma mais símples disso:

```js
let wrapper = function() {
  return func.apply(this, arguments);
};
```

> Quando um código externo chama tal `encapsulador`, ele é indistinguível da chamada da função original `func`.

Reviewer: ```suggestion
Quando um código externo chama esse `encapsulador`, ele é indistinguível da chamada da função original `func`.
```

## Emprestando um método [#method-borrowing]

> Agora vamos fazer mais uma pequena melhoria na função de hashing:

Reviewer: ```suggestion
Agora vamos fazer mais uma pequena melhoria à função de hashing:
```

Então essas duas chamadas são quase equivalentes:

```js
func.call(context, ...args);
func.apply(context, args);
```

Eles executam a mesma chamada de `func` com o contexto e argumentos dados.

Há somente uma diferença sútil em relação a `args`:

- A sintaxe spread (espalhar) `...` permite passar `args` *iterável* como a lista para `call`.
- O `apply` aceita somente `args` *parecido com array (array-like)*.

...E para objetos que são ambos iteráveis e parecidos com array, tal como um verdadeiro array, podemos usar qualquer um deles, porém `apply` irá provavelmente ser o mais rápido, porque a maioria dos interpretadores de JavaScript internamente otimizam ele melhor.

Passar todos argumentos junto com o contexto para outra função é chamado de *encaminhamento de chamada*.

Esta é a forma mais símples disso:

```js
let wrapper = function() {
  return func.apply(this, arguments);
};
```

Quando um código externo chama tal `encapsulador`, ele é indistinguível da chamada da função original `func`.

## Emprestando um método [#method-borrowing]

Agora vamos fazer mais uma pequena melhoria na função de hashing:

```js
function hash(args) {
  return args[0] + ',' + args[1];
}
```

> Até agora, ele funciona somente com dois argumentos. Seria melhor se pudesse colar qualquer número de `args`.

Reviewer: ```suggestion
Até agora, ela funciona somente com dois argumentos. Seria melhor se pudesse colar qualquer número de `args`.
```

Por quê ele funciona?

É porque o algoritmo interno do método nativo `arr.join(glue)` é muito símples.

Retirado da especificação quase "como-é":

1. Deixa a `glue` ser o primeiro argumento ou, se não houver argumentos, então uma vírgula `","`.
2. Deixa `result` ser uma string vázia.
3. Acrescentar `this[0]` ao `result`.
4. Acrescentar `glue` e `this[1]`.
5. Acrescentar `glue` e `this[2]`.
6. ...Faça isso até os itens `this.length` sejam colados.
7. Retorne `result`.

Portanto, tecnicamente ele pega o `this` e combina `this[0]`, `this[1]` ...etc tudo junto. Foi intecionalmente escrito em uma maneira que permite qualquer objeto parecido com array `this` (não é uma conscidência, vários métodos seguem esta prática). Aí está o porquê que isto também funciona com `this=arguments`.

## Decoradores e propriedades de funções

É geralmente seguro substituir uma função or um método por um decorado, exceto para uma pequena coisa. Se a função original tiver propriedades nela, tipo `func.calledCount` ou que seja, a função decorada não as forncerá. Por que ela é um encapsulador. Portanto é preciso ser cuidadoso ao usá-los.

> Alguns decoradores podem fornecer suas próprias propriedades. Exemplo, um decorador pode contar quantas vezes uma função foi invocada e quanto tempo ela levou, e expor essas informações atráves das propriedades do encapsulador.

Reviewer: ```suggestion
Alguns decoradores podem fornecer suas próprias propriedades. Exemplo, um decorador pode contar quantas vezes uma função foi invocada e quanto tempo ela levou, e expor essa informação atráves das propriedades do encapsulador.
```

> Existe uma maneira de criar decoradores que mantém o acesso às propriedades da função, mas isto requer usar um objeto `Proxy` especial para encapsular uma função. Discutiremos isso depois no artigo <info:proxy#proxy-apply>.

Reviewer: ```suggestion
Existe uma maneira de criar decoradores que mantêm o acesso às propriedades da função, mas isto requer o uso de um objeto `Proxy` especial para encapsular uma função. Discutiremos isso depois no artigo <info:proxy#proxy-apply>.
```

## Resumindo

> O *Decorador* é um encapsulador em volta de uma funçaõ que altera seu comportamento. O trabalho principal ainda é realizado pela função.

Reviewer: ```suggestion
O *Decorador* é um encapsulador à volta de uma funçaõ que altera o seu comportamento. O trabalho principal ainda é realizado pela função.
```

Decoradores podem ser vistos como "funcionalidades" ou "aspectos" que podem ser adicionados a uma função. Podemos adicionar um ou vários. E tudo isso sem alterar o seu código.

> Para implementar `cachingDecorator`, estudamos os métodos:

Reviewer: ```suggestion
Para implementar `cachingDecorator`, estudámos métodos:
```

- [func.call(context, arg1, arg2...)](mdn:js/Function/call) -- chama `func` com o seu contexto e argumentos dados.

> - [func.apply(context, args)](mdn:js/Function/apply) -- chama `func` passando `context` como `this` e um objeto `args` parecido com array dentro de uma lista de argumentos.

Reviewer: ```suggestion
- [func.apply(context, args)](mdn:js/Function/apply) -- chama `func` passando `context` como `this` e um objeto parecido com array `args` a uma lista de argumentos.
```

O *encaminhamento de chamadas* genérico é comumente realizado com `apply`:

```js
let wrapper = function() {
  return original.apply(this, arguments);
}
```

> Também vimos um exemplo de *emprestimo de método* quando pegamos um método de um objeto e `chamamos(call)` ele em um contexto de outro objeto. É bem comum pegar métodos de um array e aplicá-los aos argumentos. A alternativa é usar o resto dos parâmetros do objeto que é um array real.

Reviewer: ```suggestion
Também vimos um exemplo de *emprestimo de método* quando pegamos num método de um objeto e o chamamos (`call`) no contexto de outro objeto. É bem comum pegar em métodos de um array e aplicá-los aos argumentos. A alternativa é usar un objeto de parâmetros rest que seja um verdadeiro array.
```

> Existem muitos decoradores por aí na selva. Verifique como bem você os recebeu ao resolver as tarefas deste capítulo.

Reviewer: ```suggestion
Existem muitos decoradores por aí à volta. Verifique como bem você os percebeu ao resolver as tarefas deste capítulo.
```

hash(1, 2);
```

O truque é chamado de *emprestando método*.

> Pegamos (emprestamos) um método de combinação de um array comum (`[].join()`) e usamos `[].join.call` para executá-lo no contexto de `arguments`.

Reviewer: ```suggestion
Pegamos (emprestamos) um método de concatenação de um array comum (`[].join()`) e usamos `[].join.call` para executá-lo no contexto de `arguments`.
```

Porque ele funciona?

É porque o algoritmo interno do método nativo `arr.join(glue)` é muito símples.

Retirado da especificação quase "como-é":

1. Deixe `glue` ser o primeiro argumento ou, se não houver argumentos, então uma vírgula `","`.
2. Deixe `result` ser uma string vazia.
3. Acrescente `this[0]` ao `result`.
4. Acrescente `glue` e `this[1]`.
5. Acrescente `glue` e `this[2]`.
6. ...Faça isso até que `this.length` itens estejam colados.
7. Retorne `result`.

> Portanto, tecnicamente ele pega o `this` e combina `this[0]`, `this[1]` ...etc juntos. Foi intecionalmente escrito de uma maneira que permita qualquer objeto parecido com array `this` (não é uma conscidência, muitos métodos seguem esta prática). Aí está o porque isto também funciona com `this=arguments`.

Reviewer: ```suggestion
Portanto, tecnicamente ele pega o `this` e combina `this[0]`, `this[1]` ...etc juntos. Foi intencionalmente escrito de uma maneira que permita qualquer objeto parecido com array `this` (não é uma coincidência, muitos métodos seguem esta prática). Aí está o porque isto também funciona com `this=arguments`.
```

## Propriedades de decoradores e de funções

> É geralmente seguro substituir uma função ou um método por outro decorado, exceto por uma pequena coisa. Se a função original tiver propriedades nela, como `func.calledCount` ou qualquer outra, a função decorada não as forncerá. Por que ela é um encapsulador. Portanto é preciso ser cuidadoso ao usá-los.

Reviewer: ```suggestion
É geralmente seguro substituir uma função ou um método por outro decorados, exceto por uma pequena coisa. Se a função original tiver propriedades nela, como `func.calledCount` ou qualquer outra, a função decorada não as forncerá. Porque ela é um encapsulador. Portanto é preciso ser cuidadoso ao usá-los.
```

## Resumindo

O *Decorador* é um encapsulador à volta de uma funçaõ que altera o seu comportamento. O trabalho principal ainda é realizado pela função.

Decoradores podem ser vistos como "funcionalidades" ou "aspectos" que podem ser adicionados a uma função. Podemos adicionar um ou vários. E tudo isso sem alterar o seu código.

Para implementar `cachingDecorator`, estudámos métodos:

- [func.call(context, arg1, arg2...)](mdn:js/Function/call) -- chama `func` com o seu contexto e argumentos dados.
- [func.apply(context, args)](mdn:js/Function/apply) -- chama `func` passando `context` como `this` e um objeto parecido com array `args` a uma lista de argumentos.

O *encaminhamento de chamadas* genérico é comumente realizado com `apply`:

```js
let wrapper = function() {
  return original.apply(this, arguments);
}
```

Também vimos um exemplo de *emprestimo de método* quando pegamos num método de um objeto e o chamamos (`call`) no contexto de outro objeto. É bem comum pegar em métodos de um array e aplicá-los aos argumentos. A alternativa é usar un objeto de parâmetros rest que seja um verdadeiro array.

> Existem muitos decoradores por aí na selva. Verifique como bem você os recebeu ao resolver as tarefas deste capítulo.

Reviewer: ```suggestion
Existem muitos decoradores por aí. Verifique como bem você os percebeu ao resolver as tarefas deste capítulo.
```

Porque ele funciona?

É porque o algoritmo interno do método nativo `arr.join(glue)` é muito símples.

Retirado da especificação quase "como-é":

1. Deixe `glue` ser o primeiro argumento ou, se não houver argumentos, então uma vírgula `","`.
2. Deixe `result` ser uma string vazia.
3. Acrescente `this[0]` ao `result`.
4. Acrescente `glue` e `this[1]`.
5. Acrescente `glue` e `this[2]`.
6. ...Faça isso até que `this.length` itens estejam colados.
7. Retorne `result`.

Portanto, tecnicamente ele pega o `this` e combina `this[0]`, `this[1]` ...etc juntos. Foi intencionalmente escrito de uma maneira que permita qualquer objeto parecido com array `this` (não é uma coincidência, muitos métodos seguem esta prática). Aí está o porque isto também funciona com `this=arguments`.

## Propriedades de decoradores e de funções

É geralmente seguro substituir uma função ou um método por outro decorados, exceto por uma pequena coisa. Se a função original tiver propriedades nela, como `func.calledCount` ou qualquer outra, a função decorada não as forncerá. Porque ela é um encapsulador. Portanto é preciso ser cuidadoso ao usá-los.

> E.g. in the example above if `slow` function had any properties on it, then `cachingDecorator(slow)` is a wrapper without them.

Reviewer: ```suggestion
No exemplo acima, se a função `slow` tiver algumas propriedades, então `cachingDecorator(slow)` é um encapsulador sem elas.
```
